### ISSUE TITLE
default .uaerc in Lakka

### ISSUE THREAD

**Issue by CRP** [libretro/libretro-uae#132]
Hi, the core works well if I specify a full .uae file with all settings in it.
I would rather have a global .uaerc with main options, and then use .uae files only for variants and floppies.
But where do I put the .uaerc file? I tried everything (except root, which is not writable), but the core will not read it.

folders I tried:
/tmp/system
/tmp/system/uae-data
/storage
/storage/.config
/storage/.config/retroarch

thanks

**Comment by sonninnos** [comment]
Core options are doing that job. Any uae-conf used as a launch parameter will be appended to that core option base and will replace all duplicate settings. Observe the temporary `puae_libretro.uae` in the saves-directory.

**Comment by CRP** [comment]
So what this means is that those settings which are exposed in the gui under Core->Options are used to create `puae_libretr.uae` on the fly, correct?
If so, how do I add settings that are not exposed in the gui? For example, how do I define a path to a kickstart rom that can be used as default?

Thanks

**Comment by sonninnos** [comment]
On the fly as in at startup, yes.

Non existing settings are added by simply inserting the rows in a .uae-conf file.

For example running a file including only:
`kickstart_rom_file=<path>`
will only override the default kickstart.

But again I ask: Why all the extra trouble? What is wrong with the existing Kickstarts and what essential settings are missing?

**Comment by CRP** [comment]
One possible use would be to change the kickstart rom to be used for all content without having to change content specific uae-conf files one per one.
Other things that are not accessible as core options are resolution, as well as fine settings for cpu and speed (I believe only model can be set via core options).
I understand including all options available in configuration.txt as gui core options would be painful, thus I thought maybe we can have a default .uaerc file somewhere which is read in at emu start, overridden by core-options and finally overridden again by content specific uae-conf files.

**Comment by MajorWilly** [comment]
Sorry for butting in on this thread. :smile: 

The chosen preset model in the Core options already uses the same kickstart rom for all content until you change it to a different preset model.

Resolution is already handled by the LowRes/HiRes core option and the Zoom mode which can be used to fill the screen as needed (much much better than UAE4Arm and Amiberry IMO.)

CPU speed would be good as it currently only has 'Real' and 'Max'. When using WHDLoad with a 68020 a lot of games play too fast but can be fixed by using 'Cycle Exact' but this requires a lot of CPU power. If we could change the CPU speed to 7mz when using 68020 this fixes the fast speed without needing Cycle Exact. But probably overcomplicates things ?

Apologies if I misunderstood you.

**Comment by sonninnos** [comment]
Looks like the temporary `puae_libretro.uae` is hacked to already be that default .uaerc, so no wonder it is not read anywhere. 

I'm not at all interested in including more archaic stuff, since the goal is to have core options to do everything needed. Core options can be saved per content in RA, so that makes juggling with manual confs even more pointless. And since the scope of this core is not to be a full emulator, but rather be focused on minimally annoying way of playing games.

But of course I'm not stopping anyone from trying.

And I did already try to make more sense of those cpu speed and throttle settings, and gave up due to a lousy effort vs gain ratio. But maybe `cpu_frequency` or `cpu_clock_multiplier` could be more easily harnessed, and since they may serve a purpose, new core options are warranted. Above mentioned favorite Kickstarts or resolutions surely do not (unless a specific case is presented).

@MajorWilly Please provide a few examples of those too fast games, so I have something to work with. Weird that the disk versions aren't too fast then, since WHDLoad versions include all kinds of waits to be compatible with faster machines. Perhaps those could be fixed with some tooltypes/parameters, like nocache etc..?

**Comment by MajorWilly** [comment]
They work fine if you use WHDLoad with a 68000 but if you use 68020 there seem to be quite a few that play way too fast unless you enable Cycle Exact. As I noted above, they seem to play at roughly the correct speed with the option `cpu_throttle=-800.0` on 68020 (that's WinUAE, I don't know if PUAE has this ?) There was a thead or website that had a list with some that play too fast but I can't remember where it was. I remember these -

Barbarian (Palace)
Harlequin
Jaguar XJ220
Technocop
Magic Pockets

**Comment by sonninnos** [comment]
Nope, there is no DMA cycle stuff in here. Only exacts are cpu cycle and blitter, and `cycle_exact` enables both of them.

But regarding that speed thing: Have you tried enabling `waiting_blits`?

Here it is actually supposed to be enabled by default when: cpu is 68020 or faster, cpu_cycle_exact is disabled and immediate_blits is disabled, but it isn't enabled, because reasons..

**Comment by MajorWilly** [comment]
I've tried that option and it makes no difference. I'm not even sure what use that option is for anyway, i've never had to use it ? This doesn't really matter anyway, please use your time for more important things :smile:

**Comment by sonninnos** [comment]
WinUAE describes `waiting_blits` as 

> Compatibility hack for programs that don't wait for the blitter correctly, causing graphics corruption if CPU is too fast

Indeed it does not help with the Jaguar hyper-speed, but it does help with Nebulus 2 on A1200.
The disk version of Jaguar on A1200 is too fast even with `cycle_exact` enabled, but disabled with some `cpu_throttle` seems to do it.

Realtime `cpu_throttle` + `cycle_exact` core options incoming after some more quality assurance.
Looks like it'll also make that cpu speed setting pretty much obsolete.

**Comment by MajorWilly** [comment]
The IPF version of Jaguar has Manual Copy Protection. Maybe that's why it crapped out ?

**Comment by sonninnos** [comment]
Hah, of course, sounds reasonable.

**Comment by CRP** [comment]
Hi, I just had some spare time to do some testing.
I placed correct (both filename and md5) kickstart files in the system folder.
If I start the core from "Load Core", it starts correctly, applies the correct config (which is reflected in the savefiles/puae_libretro.uae file), and all content works correctly.
If I use a game-specific .uae file, then all core options are ignored. In other words, if my .uae file only contains the floppy0=file.adf statement, then that is the only setting which is correctly applied. The core starts with the GUI, and all settings are "empty", or set to default.
In other words, if I wish to use a custom .uae file, then that has to contain all desired settings, as it does not inherit the core options.
Is this the expected behaviour?

**Comment by sonninnos** [comment]
The core hasn't had the GUI for a long time, so therefore you have a version which also does not have the core option inheritance.

**Comment by CRP** [comment]
I am using the core which comes with the latest version of Lakka, will check how I can upgrade that.

thanks

**Comment by gouchi** [comment]
Hi,

We are using version from [commit 943e1f2](https://github.com/libretro/Lakka-LibreELEC/blob/master/packages/libretro/puae/package.mk#L22).

The time we update it, you may try to add an updated core to `/storage/cores`. Please bear in mind cores from the buildbot may have incompatibility issues with Lakka.

Thank you.

**Comment by rsn8887** [comment]
That commit is totally outdated. Please don’t post issues using Lakka, unless Lakka gets updated to latest core version, otherwise you are wasting everybody’s time, thanks.

**Comment by gouchi** [comment]
@CRP please close this issue and open an issue on [Lakka issue tracker](https://github.com/libretro/Lakka-LibreELEC/issues).

@rsn8887 [Lakka doesn’t expose RetroArch Online Updater (buildbot). Why?](https://www.lakka.tv/doc/faq/#lakka-doesnt-expose-retroarch-online-updater-buildbot-why).
